Handles redirect route.

saralein opened the pull request.

saralein committed:
Handles redirect route.

damonkelley commented on src/main/java/com/saralein/server/Constants.java:
```diff
@@ -11,6 +11,7 @@ private Constants() {}
 
     public static final Map<Integer, String> STATUS_CODES = Collections.unmodifiableMap(new HashMap<Integer, String>() {{
```
Pretty soon I think we will want move the status codes into its own class. You might look into using `Enums` for that.

damonkelley commented on src/main/java/com/saralein/server/response/Header.java:
```diff
@@ -1,30 +1,49 @@
     }
 
-    String createContents() {
+    public byte[] convertToBytes() {
```
Do you think the header should know about converting itself to bytes?

damonkelley commented on src/main/java/com/saralein/server/Controller/FileController.java:
```diff
@@ -0,0 +1,45 @@
+import com.saralein.server.response.ResponseBuilder;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
```
Based on experience, I would recommend that you try to use either `java.io` or `java.nio` for file operations. (I prefer `java.nio`)

damonkelley commented on src/main/java/com/saralein/server/Controller/FileController.java:
```diff
@@ -0,0 +1,45 @@
+        this.fileHelper = fileHelper;
+    }
+
+    public byte[] createResponse() {
```
This is related to a previous comment.

Do you think that a `Controller` should be responsible for converting a response to bytes?

damonkelley commented on src/main/java/com/saralein/server/Controller/FileController.java:
```diff
@@ -0,0 +1,45 @@
+                    .addHeader("Content-Type", getMimeType())
+                    .addBody(createBody())
+                    .build()
+                    .convertToBytes();
```
Do you think that a `Response` should have to know how to convert itself to bytes?

damonkelley commented on src/main/java/com/saralein/server/response/Header.java:
```diff
@@ -1,30 +1,49 @@
+        return  version + " " + STATUS_CODES.get(code);
+    }
+
+    private String createContents() {
```
There is a lot going on here. Maybe you could extract some methods with intention-revealing names?

damonkelley commented on src/main/java/com/saralein/server/router/SetupRoutes.java:
```diff
@@ -0,0 +1,13 @@
+
+import com.saralein.server.Controller.RedirectController;
+
+public class SetupRoutes {
```
What do you think about passing in the routes from `main`?

damonkelley commented:
What I getting at with the comments about bytes is that feels like a protocol level object shouldn't be concerned with implementation details. HTTP is a plain text protocol. It just so happens, based on a number of different factors, you need to convert that text into bytes in order to move it from point a to b.

damonkelley commented:
Of course, at some point, something needs to be responsible for converting the response into bytes

damonkelley commented:
Aside from all the bytes stuff, every looks good!

saralein committed:
Refactors Response passing, switches to java.nio, creates status code enum.

saralein commented:
Reworked the code some.  Switched to java nio where possible.  Updated a few classes so a Response is returned vs. byte array.  Created a response serializer in the connection handler, so a Response can pass all the way back and then be converted.  Created a status code enum.  Not entirely satisfied that the Constants class is still hanging out with `CRLF`, but will take a look at that in the future.

As for passing the routes in `main`, I wanted to keep a bit of separation between the cobspec routes and the bulk of the server.  I did debate passing in the routes vs. a router configuration class which sets up a data structure on the router.

zdavy commented on src/main/java/com/saralein/server/router/ServerRouter.java:
```diff
@@ -1,40 +1,49 @@
         } else {
-            return new FileResponse(request, resource, fileHelper).createResponse();
+            return new FileController(request, resource, fileHelper).createResponse();
         }
```
Is there any reason why these controllers can't conform to the same interface as the other controllers? Like can they just be injected into that list of routes?

saralein replied:
Yes, I can do that.  I might do it on the current branch I'm working on though since I'm working around the same area a lot.

zdavy replied:
OK
